### PR TITLE
Base api: jme3-alloc base api

### LINKED ISSUE #6: [Core] Provide a base interface API for jMonkeyEngine
### The base api should have the following functionalities:
- [x] Direct memory allocation.
- [x] Direct memory de-allocation.
- [x] #19
- [x] JNI Utilities.
- [x] Java Documentation.
- [x] Native code Documentation.

### PR DESCRIPTION
**Pull request by pavly-gerges**
### This PR adds the following: 
- [x] Native base api `jmealloc.h` utilities.
- [x] JVM side api `NativeBufferUtils.java` and `NativeBufferAllocator.java`.
- [x] Native image extractor `NativeBinaryLoader.java` that loads the system specific library.
- [x] Build script `{java.home}` supply to include the native headers on compile-time.
- [x] Build script enhancements including `{stdin, stderr}` output logs. 
- [x] Added eclipse temurin Open-JDK-17 to the github runner images.
- [x] Documentation for the native side api.
- [x] Documentation for the java side api.

### CODE REVIEW

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [inline jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: It will be nice to document that it uses: `{@link NativeBufferUtils#malloc(long)}`

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [inline jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: >{@link NativeBufferUtils#malloc(long)}

Should it be calloc?

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [inline jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: I wonder why this method is needed, can't we do this from java side?

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [inline jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: I wonder why this method is needed, can't we do this from java?

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [inline jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: I wonder why this method is needed, can't we do this from java?

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [inline jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: I wonder why this method is needed, can't we do this from java?

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: This is the jvm version for the GNU stdlib `void* malloc(size_t)` function, but i am trying to follow the java naming conventions (google java coding style).....let me know if you would like to break this rule and use the GNU style here.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: I want to build a full api, i am using GNU libc and the nio jni documentation, so i am still not sure if this is actually functional in jni.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: Note, I do not mean to rename the method, just was pointing to add a Javadoc on this method and mention that underlying it uses stdlib `malloc`.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: Internally on GNU libc (stdlib), calloc calls `malloc` and `memset` the buffer to zero or NULL (if it's not), i am trying to stick to the GNU libc documentation.

All other versions of memory allocations on the stdlib uses malloc internally.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: Yeah, you are right, i wasn't aware of this, thank you, i am going to remove this.

Simply call from java `Buffer#capacity()` will return the number of elements the buffer holds.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: This is still a WIP, finding the size of the whole buffer in one line from C may be better, but ofc you can do this from java by multiplying the capacity by the buffer type size which is a byte in this case.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: Oh okay, i misunderstood this sorry.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: I am not sure if these will work on java nio buffers: 
1) `memoryMove`
2) `memorySet`
3) `realloc`
4) `reallocarray`

Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: I will test them today, and we should decide here before merging this PR.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: Done

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [inline jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: >{@link NativeBufferUtils#malloc(long)}

I could not find function `malloc(long)` on NativeBufferUtils, do you mean `{@link NativeBufferUtils#memoryAlloc(long)}`?

Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: Yeah, thanks for pointing out.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: This is achievable through java, i have removed these native methods.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: I will provide a testcase for this method, it provides the capability to set all the buffer byte elements with a [value], it can be done through java with a for-loop, but i prefer to keep the native implementation anyway, the native implementation uses the memory address and the size to define the buffer boundaries.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: Removed `realloc` and `reallocarray`, since the `java.nio.Buffer` has a constant capacity value, so those have an undefined behavior, while `memorySet` and `memoryMove` are working fine, i will provide techdemos for them....

**Comment by pavly-gerges** [comment]
@Ali-RS I am done laying out the base api and the documentation, please give them a look before merging.

- About other methods `memorySet()`, `memoryMove()`, and `memoryCopy()` there will be techdemos showing their usages.
- As for `TestLibrary.java` class it was for prototyping and the API shouldn't have a main runnable method, and the `jme3-testalloc` will have a list of techdemos later on.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [inline jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: >Note: Maximum size is 32-bit for 32-bit architectures and 64-bit for 64-bit architectures.

Probably better to change it to:

>Note: Maximum buffer size can be 2^32 bytes for 32-bit architectures and 2^64 bytes for 64-bit architectures. 

what do you think?

Comment (Ali-RS) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java:69]: or do you mean the max size can be 2^32 (for 32-bit CPU) and 2^64 (for 64-bit CPU)?

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [inline jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java:95]: Usually, `from` should come first and `to` next.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [inline jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java:105]: Usually, `from` should come first and `to` next.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java:95]: The [native method](https://github.com/Software-Hardware-Codesign/jme-alloc/blob/28ae779618ba64293c65cfbf6c8376afe71182fe/jme3-alloc-native/src/include/jmealloc.h#L92-L96) starts with the [to] buffer, i would like to keep it aligned to the native method.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java:105]: The native method starts with the [to] buffer.

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java:95]: Yeah, I mean to update both of them if it is ok:)

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: I cannot recall what is `2` in `2^64`

I was documenting the ability to allocate the size based on the parameter provided, which is long, long is 64-bit number but it's a 32-bit number on 32-bit binaries.....

But reading this again, i am wrong, the parameter is in `bytes` units, so it should be 32 or 64 bytes (and not bits).

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java:95]: I mean the native method `memmove(void* to, void* from, size_t size)` from stdlib starts with the [to] buffer.

Comment (Ali-RS) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java:95]: Oh, I see. Sorry!

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (Ali-RS) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: >I cannot recall what is 2 in 2^64

Long.MAX_VALUE which actually is (2^63) - 1 considering the sign bit.

So it should be:

>Note: Maximum buffer size can be Integer.MAX_VALUE bytes for 32-bit architectures and Long.MAX_VALUE bytes for 64-bit architectures.

Comment (Ali-RS) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: By the way, afaik there is no such limitation for `long` on java for 32-bit systems. Is this limitation on C code? 

Please take a look at this question: https://stackoverflow.com/questions/50996253/in-c-what-is-the-size-of-a-long-for-a-32-bit-machine-and-for-a-64-bit-machine

Maybe you should remove that note as the limit is not platform specific but depends on C implementation as mentioned in the above link.

Comment (Ali-RS) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: And by the way, JME uses an `int` for size:

https://github.com/jMonkeyEngine/jmonkeyengine/blob/4db02fae9dd8ebb213b3a06316a53019ef9b620f/jme3-core/src/main/java/com/jme3/util/BufferAllocator.java#L23

Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: > Is this limitation on C code?

In C, `size_t` is defined as `unsigned long int` which is architecture dependent and it's equivalent to `unsigned long` (the gnu docs), 32-bit (or 4 bytes) for 32 architectures and 64-bit (or 8 bytes) for 64 architectures.

In java, `jlong` is defined as `__int64` on windows and `long long` on linux, `long long` is 64-bit despite the system architecture, while native `long` is architecture dependent.

But, there is a note on windows `jni_md.h`:
```
// 'long' is always 32 bit on windows so this matches what jdk expects
```
So, here we don't know the architecture of the system from java, if it's a 32-bit binary the `size_t` max number will be a 32-bit number (or 4 bytes or 0x(FF)(FF)(FF)(FF)), if it was a 64-bit binary, the `size_t` will be a 64-bit number....so we have to always use a 64-bit number anyway on the java side.

Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: I have not tested passing a 64-bit jlong `Long.MAX_VALUE` to a native `long` parameter on a 32-bit system.....i guess it will lead to overflow and the number will be reset to (0) because `unsigned long int` starts from 0 up-to 2^32 on a 32-bit binary, because the native method accepts native `unsigned long` which is architecture dependent, so we should document this properly.....that the native method is architecture dependent, if you have any tips to give on documentation of this part, please let me know.

Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: > So it should be:
> 
> > Note: Maximum buffer size can be Integer.MAX_VALUE bytes for 32-bit architectures and Long.MAX_VALUE bytes for 64-bit architectures.

I think this might be the best description for this docs part, this code runs internally a native code which is architecture specific.

Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: @Ali-RS okay, i have refactored this, please review the new documentation.

Comment (Ali-RS) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: Looks good to me, thanks

**Review by Ali-RS — APPROVED** [review]

File: jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java
Comment (pavly-gerges) [reply jme3-alloc/src/main/java/com/jme3/alloc/util/NativeBufferUtils.java]: Great, thanks for your contributions, going to merge this soon.

**Pull request merged**
